waiter/src/waiter/scheduler.clj

```diff
@@ -642,3 +646,182 @@
           "WAITER_PASSWORD" (service-id->password-fn service-id)
           "WAITER_SERVICE_ID" service-id
           "WAITER_USERNAME" "waiter"}))
+
+;; Support for tracking service instance launching time stats
+
+(def instance-counts-zero {:requested 0 :running 0})
+
+(defn- make-launch-tracker
+  ([service-id]
+   (make-launch-tracker service-id instance-counts-zero #{}))
+  ([service-id instance-counts known-instance-ids]
+   {:instance-counts instance-counts
+    :instance-scheduling-start-times []
+    :known-instance-ids known-instance-ids
+    :service-schedule-timer (metrics/service-timer service-id "launch-overhead" "schedule-time")
+    :service-startup-timer (metrics/service-timer service-id "launch-overhead" "startup-time")
+    :starting-instance-ids []}))
+
+(defn update-launch-trackers
+  "Track metrics on launching service instances, specifically:
+   1) the time from requesting a new instance until it is scheduled (the \"schedule\" time), and
```

Comment:
We discussed this in our meeting today, and the resolution was that we'd rather have the simplicity of the current implementation over the finer-grained accuracy of the solution you propose; i.e., since we don't mind having some coarser error margins, we think the current passive-observer method of tracking these metrics is sufficiently accurate for our needs.